homeassistant/components/http/session.py

```diff
@@ -0,0 +1,127 @@
+SESSION_CACHE_SIZE = 16
+
+
+def _get_cookie_name(is_secure: bool) -> str:
+    """Return the cookie name."""
+    return PREFIXED_COOKIE_NAME if is_secure else COOKIE_NAME
+
+
+class HomeAssistantCookieStorage(EncryptedCookieStorage):
+    """Home Assistant cookie storage."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the cookie storage."""
+        super().__init__(
+            hass.auth.session.key,
+            cookie_name=COOKIE_NAME,
+            max_age=int(REFRESH_TOKEN_EXPIRATION),
+            httponly=True,
+            samesite="Lax",
+            encoder=json_dumps,
+            decoder=json_loads,
+        )
+        self._hass = hass
+
+    def _secure_connection(self, request: Request) -> bool:
+        """Return if the connection is secure (https)."""
+        return is_cloud_connection(self._hass) or request.secure
+
+    def load_cookie(self, request: Request) -> str | None:
+        """Load cookie."""
+        is_secure = self._secure_connection(request)
+        cookie_name = _get_cookie_name(is_secure)
+        self.cookie_params["secure"] = is_secure
+        return request.cookies.get(cookie_name)
+
+    @lru_cache(maxsize=SESSION_CACHE_SIZE)
+    def _decrypt_cookie(self, cookie: str) -> Session:
+        """Decrypt and validate cookie."""
+        try:
+            data = self._decoder(
+                self._fernet.decrypt(cookie.encode("utf-8"), ttl=self.max_age).decode(
+                    "utf-8"
+                )
+            )
+            return Session(None, data=data, new=False, max_age=self.max_age)
+        except (InvalidToken, *JSON_DECODE_EXCEPTIONS):
+            _LOGGER.warning(
+                "Cannot decrypt/parse cookie value, create a new fresh session"
+            )
+            return Session(None, data=None, new=True, max_age=self.max_age)
+
+    async def load_session(self, request: Request) -> Session:
+        """Load session."""
+        # Split parent function to use lru_cache
+        cookie = self.load_cookie(request)
+        if cookie is None:
+            return Session(None, data=None, new=True, max_age=self.max_age)
+        return self._decrypt_cookie(cookie)
+
+    async def save_session(
+        self, request: Request, response: StreamResponse, session: Session
+    ) -> None:
+        """Save session."""
+
+        is_secure = self._secure_connection(request)
+        cookie_name = _get_cookie_name(is_secure)
+
+        if session.empty:
+            response.del_cookie(cookie_name)
+        else:
+            params = self.cookie_params.copy()
+            params["secure"] = is_secure
+            params["max_age"] = session.max_age
+
+            cookie_data = self._encoder(self._get_session_data(session)).encode("utf-8")
+            response.set_cookie(
+                cookie_name,
+                self._fernet.encrypt(cookie_data).decode("utf-8"),
+                **params,
+            )
```

Comment:
Can any of the method calls to create and set the cookie raise an exception?

Reply:
encrypt will raise a TypeError if `cookie_data` is not `bytes`, but the encoder is returning a str, which will be encoded. So the TypeError is not raised.

The encoder itself could raise `JSON_ENCODE_EXCEPTIONS` so maybe we could catch these. Normally these errors shouldn't be raised as we control the data, which is encoded.
What should happen if an Error is raised?

Reply:
If we control the data it's ok I think. It's important that we don't leave any room for an attack that can crash the server.